### PR TITLE
Implement `reselect` recommendations and best practices

### PR DESCRIPTION
**Pull request by joeyparis**
This implements the `reselect` recommendations in [this comment](https://github.com/react-boilerplate/react-boilerplate/issues/2806#issuecomment-554855162). It is one of the **To Do**s on #2812.

I believe I have fully refactored the usage of `reselect` in the containers, templates, generators, and tests per @markerikson's recommendations. Test coverage and success remains at 100%, although I will be honest I am relatively new at test suites, and someone should probably double-check my work.

## React Boilerplate

Thank you for contributing! Please take a moment to review our [**contributing guidelines**](https://github.com/react-boilerplate/react-boilerplate/blob/master/CONTRIBUTING.md)
to make the process easy and effective for everyone involved.

**Please open an issue** before embarking on any significant pull request, especially those that
add a new library or change existing tests, otherwise you risk spending a lot of time working
on something that might not end up being merged into the project.

Before opening a pull request, please ensure:

- [✔️] You have followed our [**contributing guidelines**](https://github.com/react-boilerplate/react-boilerplate/blob/master/CONTRIBUTING.md)
- [❌] Double-check your branch is based on `dev` and targets `dev`
-- _Not relevant for the nature of this pull request_
- [✔️] Pull request has tests (we are going for 100% coverage!)
- [✔️] Code is well-commented, linted and follows project conventions
- [✔️] Documentation is updated (if necessary)
- [✔️] Internal code generators and templates are updated (if necessary)
- [✔️] Description explains the issue/use-case resolved and auto-closes related issues

Be kind to code reviewers, please try to keep pull requests as small and focused as possible :)

**IMPORTANT**: By submitting a patch, you agree to allow the project
owners to license your work under the terms of the [MIT License](https://github.com/react-boilerplate/react-boilerplate/blob/master/LICENSE.md).

### CODE REVIEW

File: app/containers/HomePage/index.js
Comment (joeyparis) [inline app/containers/HomePage/index.js]: I opted for removing `createStructuredSelector` because of the consideration for removing `reselect` as a dependency. It does lead to using 4 separate `useSelector` hooks rather than one deconstructed one. I'm not sure if this is desirable or not, so I'm definitely up for discussing other alternatives!

**Comment by julienben** [comment]
Thanks a lot for implementing this @joeyparis! At first glance, it looks like exactly what was needed. Will review in full tomorrow. Also, we might want to merge this into `dev` directly after #2812 as the squash merge will overwrite your contribution history.

File: app/containers/App/selectors.js
Comment (justingreenberg) [inline app/containers/App/selectors.js]: hi @joeyparis - is there a reason you removed the closures?

we wrap selectors in factories so that each selector instance maintains a private cache for memoization when using across different components, see:

https://github.com/reduxjs/reselect#sharing-selectors-with-props-across-multiple-component-instances

Comment (joeyparis) [reply app/containers/App/selectors.js]: Hi @justingreenberg, I was really strictly following [this comment](https://github.com/react-boilerplate/react-boilerplate/issues/2806#issuecomment-554855162), which was one of the **To Do**'s. So honestly I can't say I put any deep thought into the use of factories or lack thereof. If you feel there's an alternative way to use factories, and align with the to-do/comment I'm all ears!

Comment (markerikson) [reply app/containers/App/selectors.js]: Right, and per my comments, the existing code is doing that unnecessarily.  There's no reason to have multiple selector instances all reading the exact same single value.  The factory pattern is only needed when you will be calling the "same" selector with varying inputs, such as `const todo = selectTodoById(state, ownProps.id)`.  In that case, the inconsistent parameters would cause a single selector instance to not memoize at all, so you need separate instances (usually per component).

I talked about this in more detail here:

https://blog.isquaredsoftware.com/2017/12/idiomatic-redux-using-reselect-selectors/

**Review by joeyparis — COMMENTED** [review]
> Thanks a lot for implementing this @joeyparis! At first glance, it looks like exactly what was needed. Will review in full tomorrow. Also, we might want to merge this into `dev` directly after #2812 as the squash merge will overwrite your contribution history.

Awesome, sounds like a plan! Just let me know if I need to change which branch it's being merged into.

**Comment by julienben** [comment]
To me this looks good.

@justingreenberg Do you still have objections or are you satisfied by the explanation given by @markerikson? Will wait for your reply to do anything about this branch.

And in any case, this will go directly into `dev` only after the Redux Toolkit switch is done. (I'm making progress on the documentation updates.)

**Review by vineyardbovines — APPROVED** [review]
This LGTM as well

**Comment by justingreenberg** [comment]
@joeyparis thanks for clarifying 👍🏼 @julienben yes LGTM! 😊

**Comment by joeyparis** [comment]
Do I need to do anything at this time? Or are we just waiting until the Redux Toolkit switch is done? Thanks for the feedback everyone!

**Comment by julienben** [comment]
Nope, nothing to do I believe.

**Comment by Can-Sahin** [comment]
In `app/containers/ReposManager/selectors.js` did you forget to import `createSelector` from `@reduxjs/toolkit` ? Because `reselect` isn't need anymore

```javascript
// import { createSelector } from 'reselect'; // INSTEAD
import { createSelector } from '@reduxjs/toolkit'; 

```

**Comment by joeyparis** [comment]
@Can-Sahin looks like I did. Fixed and updated!

File: .all-contributorsrc
Comment (pedro-victor) [inline .all-contributorsrc]: I believe this modification breaks the file.

File: app/containers/HomePage/index.js
Comment (pedro-victor) [reply app/containers/HomePage/index.js]: I think that's correct. You could also remove `reselect` from `package.json` as well.

**Review by pedro-victor — COMMENTED** [review]
Apart from the `.all-contributorsrc` and removing `reselect` from `package.json`, I think this PR does basically what I have been using on my own projects (which I mistakenly pushed #2921 ) and looks to be in line with the recommendations.

**Comment by Can-Sahin** [comment]
Merging due to the major v5 release. Read the discussion [here](https://github.com/react-boilerplate/react-boilerplate/issues/2906#issuecomment-653513532)

**Comment by Can-Sahin** [comment]
Edit: Sorry for merging this somewhere else due to #2935. If you want please re submit this PR to the dev branch if that means anything to you

**Pull request merged**
